Add internal account creation

tristanvuong2021 opened the pull request.

SanjayVas reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302)* status: 0 of 38 files reviewed, 1 unresolved discussion (waiting on @SanjayVas and @tristanvuong2021)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/server/SpannerKingdomDataServer.kt, line 21 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MkrrPCI0RpEV2Ab8QFA:-MkrrPCJ2ckQNjTQt1ca:b-qqu1ot) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/db301d290d97e12a4dd57438af03f015d09acb8d/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/server/SpannerKingdomDataServer.kt#L21)):*
> ```Kotlin
> import org.wfanet.measurement.common.commandLineMain
> import org.wfanet.measurement.common.identity.RandomIdGenerator
> import org.wfanet.measurement.common.identity.RandomStringGenerator
> ```

We don't need a separate generator for random strings. All random values are just numbers, and any number can be encoded to a string. ExternalId's `apiId` property is an example of this.

The token should really be an ExternalId, i.e. a number everywhere internally and encoded as a string by the calling layer (e.g. by some util for account creation).


<!-- Sent from Reviewable.io -->

tristanvuong2021 commented:
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/server/SpannerKingdomDataServer.kt, line 21 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MkrrPCI0RpEV2Ab8QFA:-MksXXbU2q8jte_d5jZL:b-gjh9hg) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/42e60d1879e7afc0cc17df57c66924fea863137c/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/server/SpannerKingdomDataServer.kt#L21)):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

We don't need a separate generator for random strings. All random values are just numbers, and any number can be encoded to a string. ExternalId's `apiId` property is an example of this.

The token should really be an ExternalId, i.e. a number everywhere internally and encoded as a string by the calling layer (e.g. by some util for account creation).
</blockquote></details>

I removed the string generator and changed activation token and measurement consumer creation token to a number internally, although I kept the name activation token and measurement consumer creation token.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> Reviewed 1 of 38 files at r1, 11 of 25 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302)* status: 12 of 38 files reviewed, 5 unresolved discussions (waiting on @SanjayVas and @tristanvuong2021)

___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl, line 83 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBF69pB_UVUDF7sq-W:-MlBF69pB_UVUDF7sq-X:b-d6i9xf) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fe5e29eefc49b7b8553a3f6a9fc660b46ab5dc3/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl#L83)):*
> ```
>   CreateTime TIMESTAMP NOT NULL OPTIONS (allow_commit_timestamp = true),
>   UpdateTime TIMESTAMP NOT NULL OPTIONS (allow_commit_timestamp = true),
> ) PRIMARY KEY (AccountId);
> ```

Add foreign key constraints. At the very least, OwnedMeasurementConsumerId should reference MeasurementConsumers. I'm not sure whether foreign keys for the same table are supported, but if so CreatorAccountId should reference Accounts.
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl, line 94 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBFp0sBul6kG1pVQUP:-MlBFp0sBul6kG1pVQUQ:b-dyjuzl) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fe5e29eefc49b7b8553a3f6a9fc660b46ab5dc3/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl#L94)):*
> ```
> Password
> ```
PasswordHash to make it clear that this isn't a password. It should also be BYTES
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl, line 95 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBFwam1ybVeG7kYYMR:-MlBFwam1ybVeG7kYYMS:b-sxkiqq) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fe5e29eefc49b7b8553a3f6a9fc660b46ab5dc3/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl#L95)):*
> ```
> STRING(MAX)
> ```
bytes
___
*[src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl, line 98 at r2](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBFX_h6c-09ZMU0hvM:-MlBFX_h6c-09ZMU0hvN:bz7hp1f) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/6fe5e29eefc49b7b8553a3f6a9fc660b46ab5dc3/src/main/kotlin/org/wfanet/measurement/kingdom/deploy/gcloud/spanner/measurement.sdl#L98)):*
> ```
> 
>   FOREIGN KEY (AccountId) REFERENCES Accounts(AccountId),
> ) PRIMARY KEY (UsernameIdentityId);
> ```

The primary key should be `(AccountId, UsernameIdentityId)` as the identity binds to a single account.

Group the PKey column declarations as well.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302)* status: 12 of 38 files reviewed, 6 unresolved discussions (waiting on @SanjayVas and @tristanvuong2021)

*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBKK4L09vlt5lUwvYm:-MlBKK4L09vlt5lUwvYn:b3mxhnq) (no related file):*
It may be best to get the schema and proto changes in a separate PR first so you can push changes to that in response to comments without having to first fix all the dependent code.

---


<!-- Sent from Reviewable.io -->

tristanvuong2021 commented:
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBKK4L09vlt5lUwvYm:-MlBTw6K0bjihr4ueqTF:b-ovwaw7) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

It may be best to get the schema and proto changes in a separate PR first so you can push changes to that in response to comments without having to first fix all the dependent code.
</blockquote></details>

Good idea.

---


<!-- Sent from Reviewable.io -->

tristanvuong2021 commented:
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/302#-MlBKK4L09vlt5lUwvYm:-MlC9hsHAIQ-dAaQ0Ram:b-sl8tj5) (no related file):*
<details><summary><i>Previously, tristanvuong2021 (Tristan Vuong) wrote…</i></summary><blockquote>

Good idea.
</blockquote></details>

Made a new pr: https://github.com/world-federation-of-advertisers/cross-media-measurement/pull/313

---


<!-- Sent from Reviewable.io -->
